lib/encodings/base64.fz

```diff
@@ -0,0 +1,155 @@
+# Base64 encoding and decoding as defined in RFC 4648
+public base64 is
+
+  # allows redefinition for base64url
+  module get_alphabet => "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/".as_codepoints.as_array
+  module get_char_62 u8 => 43 # +
+  module get_char_63 u8 => 47 # /
+  module get_encoding_name => "base64"
+
+  # NYI: compiler does currently not optimize features without arguments to treat them like fields
+  char_62_ascii := get_char_62
+  char_63_ascii := get_char_63
+  alphabet := get_alphabet
+  encoding_name := get_encoding_name
+
+  # Encodes a given byte sequence in base64, output length is padded to multiple of 4
+  public encode (data Sequence u8) String =>
+
+    # 0 -> 0; 1 -> 2; 2 -> 1
+    pad_size := (3 - data.count%3)%3     # length of the padding: number of padded zero bytes in input as well as number of '=' in output
+    in_len := data.count + pad_size      # length of zero padded input data
+
+    # this guarantees that input length is always multiple of 3
+    input := array u8 in_len i->if i < data.count then data[i] else 0
+
+    for b0 := 0, b0 + 3
+        b1 := 1, b1 + 3
+        b2 := 2, b2 + 3
+        res := "", res + block # the encoded input data
+    while b2 < in_len
+      # get 3 bytes (24 bits) and convert them into a number with 24 bits
+      n u32 := (input[b0].as_u32 << 16) + (input[b1].as_u32 << 8) + input[b2].as_u32
+
+      # extract four sextets from 24 bit number
+      n0 => ((n >> 18) & 63).as_i32
+      n1 => ((n >> 12) & 63).as_i32
+      n2 => ((n >> 6) & 63).as_i32
+      n3 => (n & 63).as_i32
+
+      # encode sextets
+      block := alphabet[n0] + alphabet[n1] + alphabet[n2] + alphabet[n3] # NYI: OPTIMIZATION: lots of string concatenation
+    else
+      res.substring 0 (res.byte_length - pad_size) + "=" * pad_size
+
+  # decodes a base64 string, decoding is strict as required by RFC 4648
+  # non alphabet characters, line breaks, missing padding cause errors
+  public decode_str(data String) =>
+    if data.byte_length != data.codepoint_length
```

Comment:
This if should be removed and all the logic of detecting bad input should be in `decode`.

Reply:
The logic is in `decode` already, the reason I've added this, is to generate better error messages.  `decode` works on a byte array, so when a string contains a multi byte code point, the error message will contain this code point and not some cryptic letter resulting from the first byte being converted into a code point.

Reply:
Then you should probably add a `decode_internal` that returns either a result or a position of the error in the input.
Then in `decode` and in `decode_str` build an appropriate error message.

Reply:
I'd prefer not to do this, because then I would once again have to figure out what is the error at position `i` in the input and even do so in two places (`decode` and `decode_str`).

My suggestion would be to always obtain the next `codepoint` when the error happens (performance in the error case should not matter anyways) and include it in the error message like this:
`error: invalid base64 input at byte position 3, decoding to Unicode character '🌏'`

I think this should work for ASCII/byte input too.

This is the code to get, the next codepoint (from the byte array that `encode` receives) when the error occurs 
```
inv_char := String.type.from_bytes (data.drop i)
                            .substring_codepoint 0 1
```